dali/pipeline/util/bounding_box.h

```diff
@@ -0,0 +1,167 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.c++ copyswa
+
+#ifndef DALI_PIPELINE_UTIL_BOUNDING_BOX_H_
+#define DALI_PIPELINE_UTIL_BOUNDING_BOX_H_
+
+#include <algorithm>
+#include <utility>
+
+#include "dali/error_handling.h"
+
+namespace dali {
+
+class BoundingBox {
+ public:
+  static const size_t kSize = 4;
+
+  BoundingBox(const BoundingBox& other)
+      : left_{other.left_},
+        top_{other.top_},
+        right_{other.right_},
+        bottom_{other.bottom_},
+        area_{other.area_} {}
+
+  BoundingBox(BoundingBox&& other) noexcept : BoundingBox() { swap(*this, other); }
+
+  BoundingBox& operator=(BoundingBox other) {
+    swap(*this, other);
+    return *this;
+  }
+
+  friend void swap(BoundingBox& lhs, BoundingBox& rhs) {
+    using std::swap;
+    swap(lhs.left_, rhs.left_);
+    swap(lhs.right_, rhs.right_);
+    swap(lhs.top_, rhs.top_);
+    swap(lhs.bottom_, rhs.bottom_);
+    swap(lhs.area_, rhs.area_);
+  }
+
+  static BoundingBox FromLtrb(const float* data) {
+    return FromLtrb(data[0], data[1], data[2], data[3]);
+  }
+
+  static BoundingBox FromLtrb(float l, float t, float r, float b) {
+    DALI_ENFORCE(l >= 0 && l <= 1.f,
+                 "Expected 0 <= left <= 1. Received: " + to_string(l));
+    DALI_ENFORCE(t >= 0 && t <= 1.f,
+                 "Expected 0 <= top <= 1. Received: " + to_string(t));
+    DALI_ENFORCE(r >= 0 && r <= 1.f,
+                 "Expected 0 <= right <= 1. Received: " + to_string(r));
+    DALI_ENFORCE(b >= 0 && b <= 1.f,
+                 "Expected 0 <= bottom <= 1. Received: " + to_string(b));
+
+    DALI_ENFORCE(l <= r, "Expected left <= right. Received: " + to_string(l) +
+                             " <= " + to_string(r));
+    DALI_ENFORCE(t <= b, "Expected top <= bottom. Received: " + to_string(t) +
+                             " <= " + to_string(b));
+
+    return {l, t, r, b};
+  }
+
+  static BoundingBox FromXywh(const float* data) {
+    return FromXywh(data[0], data[1], data[2], data[3]);
+  }
+
+  static BoundingBox FromXywh(float x, float y, float w, float h) {
+    DALI_ENFORCE(x >= 0 && x <= 1.f,
+                 "Expected 0 <= x <= 1. Received: " + to_string(x));
+    DALI_ENFORCE(y >= 0 && y <= 1.f,
+                 "Expected 0 <= y <= 1. Received: " + to_string(y));
+    DALI_ENFORCE(w >= 0 && w <= 1.f,
+                 "Expected 0 <= width <= 1. Received: " + to_string(w));
+    DALI_ENFORCE(h >= 0 && h <= 1.f,
+                 "Expected 0 <= height <= 1. Received: " + to_string(h));
+
+    DALI_ENFORCE(x + w <= 1, "Expected x + width <= 1. Received: " +
+                                 to_string(x) + " + " + to_string(w));
+    DALI_ENFORCE(y + h <= 1, "Expected y + height <= 1. Received: " +
+                                 to_string(y) + " + " + to_string(h));
+
+    return {x, y, x + w, y + h};
+  }
+
+  BoundingBox ClampTo(const BoundingBox& other) const {
```

Comment:
When the boxes are non-overlapping, this produces an invalid box (left > right)

Reply:
accepted